typing/typecore.ml

```diff
@@ -990,13 +968,134 @@ let rec has_literal_pattern p = match p.ppat_desc with
   | Ppat_or (p, q) ->
      has_literal_pattern p || has_literal_pattern q
 
-exception Need_backtrack
-
 let check_scope_escape loc env level ty =
   try Ctype.check_scope_escape env level ty
   with Unify trace ->
     raise(Error(loc, env, Pattern_type_clash(trace, None)))
 
+type pattern_checking_mode =
+  | Normal
+  (** We are checking user code. *)
+
+  | Counter_example of counter_example_checking_info
+  (** We are checking a counter-example to exhaustivity
+      generated by the pattern-matching analyser. *)
+
+
+(** In [Counter_example] mode the responsibility of [type_pat] is to
+   check that a counter-example to exhaustivity is valid: it has to be
+   type-correct in a slightly different sense than for user-code, see
+   [splitting_mode] below, and also capture at least one concrete
+   value.
+
+   [counter_example_checking_info] contains various parameters for
+   this mode of operation. *)
+and counter_example_checking_info = {
+    explode: explosion_fuel;
```

Comment:
I kept the name `explode` to reduce the changes in the `type_pat` codebase, and I used a custom type name to be able to consistently document each field at the place where the type is defined, rather than documenting the record field itself (I did that at first and the result was inconvenient).

Reply:
For the record (no pun intended) I like @garrigue's suggestion better.
I'm not sure reducing the diff should be the goal here. Your aim is to make the code more readable, I think like it'd be slightly better if the field was named `explosion_fuel` rather than `explode`.

(Also, I have a personal dislike of type aliases, I think that most of their use are a mistake, this particular case being an example)